system/lib/wasmfs/file.cpp

```diff
@@ -40,33 +40,180 @@ void DataFile::Handle::preloadFromJS(int index) {
 // Directory
 //
 
+void Directory::Handle::finishInsertingChild(const std::string& name,
+                                             std::shared_ptr<File> child) {
+  // Update the dcache and set the child's parent.
+  auto& dcache = getDir()->dcache;
+  auto [_, inserted] = dcache.insert({name, {DCacheKind::Normal, child}});
+  assert(inserted && "inserted child already existed!");
+  assert(child->locked().getParent() == nullptr);
+  child->locked().setParent(getDir());
+}
+
+std::shared_ptr<File> Directory::Handle::getChild(const std::string& name) {
+  // Unlinked directories must be empty, without even "." or ".."
+  if (!getParent()) {
+    return nullptr;
+  }
+  if (name == ".") {
+    return file;
+  }
+  if (name == "..") {
+    return getParent();
+  }
+  // Check whether the cache already contains this child.
+  auto& dcache = getDir()->dcache;
+  if (auto it = dcache.find(name); it != dcache.end()) {
+    return it->second.file;
+  }
+  // Otherwise check whether the backend knows about the child.
+  auto child = getDir()->getChild(name);
+  if (!child) {
+    return nullptr;
+  }
+  finishInsertingChild(name, child);
```

Comment:
This surprises me for similar reasons as in the previous PR: we call `getDir()->getChild()`, which seems like it should be a `const` method, but from this code it seems like it has a side effect and that side effect will start to insert a child, but not finish doing so. Is that right?

If so perhaps `getDir()->getChild()` could be `getOrAddChild` or such? But given that the `dcache` appears to be used in this method before that call, there is no chance the child already exists, so perhaps `getDir()->getChild()` is only called to actually create a new child? If so perhaps it could be `createChild`?

Reply:
`Directory::getChild` never creates a new child file, but it will create a new child `File` if the underlying file already exists. That new `File` needs to have its metadata (e.g. parent pointer and presence in the dcache) updated before it can be returned to the caller of `Directory::Handle::getChild`.

Perhaps `finishInsertingChild` should instead be `finishInitializingChild` or `updateChildMetadata`?

Reply:
Could it be `insertChild`? I'm not sure what part of inserting it is handled beforehand, as it sounds it does everything. That is, what else is there aside from adding the pointer from the child to the parent (and in reverse) and also adding it to the dcache?